### PR TITLE
feat: add allowUnsafeSSL config for self-signed certificates

### LINKED ISSUE #304: support github enterprise with self-signed certificate
My company use github enterprise with a self-signed certificate. Every time I want execute any commands of this extension to do some operation on github enterprise, it will throw 

```
Error: request to https://github.intra.example.com/api/v3/repos/huanshuai/repo failed, reason: self signed certificate in certificate chain
```

like error and failed.

I think it may be due to bug similar to bitinn/node-fetch#19, the solution could be creating a workspace config item for whether reject self-signed certificate or not. If the solution is ok, I could pr for this (maybe days) later

### PR DESCRIPTION
**Pull request by lihuanshuai**
resolve #304 

Set `github.allowUnsafeSSL` to `true` in workspace settings to overcome https self-signed certificates problem.

### CODE REVIEW

**Comment by KnisterPeter** [comment]
@lihuanshuai Sorry for late reply. Your code is not compiling right now. Please see the log in travis for the compile error.

File: package.json [hunk 1]
```diff
@@ -68,6 +68,12 @@
           "default": false,
           "scope": "resource"
         },
+        "github.rejectUnauthorized": {
```
Comment (KnisterPeter) [inline package.json]: I would like to name this `allowUnsafeSSL`  and default it to `false`.
With that name its much more clear what it does.

File: src/provider/client.ts [hunk 2]
```diff
@@ -18,7 +19,9 @@ export async function createClient(git: Git, tokens: Tokens, uri: vscode.Uri,
   }
   switch (tokenInfo.provider) {
     case 'github':
-      return new GithubClient(protocol, hostname, tokens[hostname].token, logger);
+      return new GithubClient(
```
Comment (KnisterPeter) [inline src/provider/client.ts]: Please add it for the gitlab client as well.

File: src/provider/github/client.ts [hunk 3]
```diff
@@ -14,8 +14,9 @@ export class GithubClient implements Client {
 
   public name = 'GitHub Client';
 
-  constructor(protocol: string, hostname: string, token: string, logger: (message: string) => void) {
-    this.client = getClient(this.getApiEndpoint(protocol, hostname), token, logger);
+  constructor(protocol: string, hostname: string, token: string, logger: (message: string) => void,
+              rejectUnauthoized = false) {
```
Comment (KnisterPeter) [inline src/provider/github/client.ts]: This should default to `true` right now I think.
Default default should be to reject invalid ssl chains.

**Comment by lihuanshuai** [comment]
@KnisterPeter sorry for late reply. The compiling problem has confused me a lot. I am not quite familar with typing system of typescript. although the error reported, The compiled javascript code worked on my mac.

`request` is a `RequestInit` instance. In `node_modules/@types/node-fetch/index.d.ts`, `RequestInit` is defined as:

```typescript
interface RequestInit {
	//whatwg/fetch standard options
	method?: string;
	headers?: HeaderInit | { [index: string]: string };
	body?: BodyInit;
	redirect?: RequestRedirect;

	//node-fetch extensions
	timeout?: number; //=0 req/res timeout in ms, it resets on redirect. 0 to disable (OS limit applies)
	compress?: boolean; //=true support gzip/deflate content encoding. false to disable
	size?: number; //=0 maximum response body size in bytes. 0 to disable
	agent?: Agent; //=null http.Agent instance, allows custom proxy, certificate etc.
	follow?: number; //=20 maximum redirect count. 0 to not follow redirect

	//node-fetch does not support mode, cache or credentials options
}
```
`agent` field should be ok. but In `/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/lib.dom.d.ts`, which is actually used, `RequestInit` is defined as:

```typescript
interface RequestInit {
    signal?: AbortSignal;
    body?: Blob | BufferSource | FormData | string | null;
    cache?: RequestCache;
    credentials?: RequestCredentials;
    headers?: HeadersInit;
    integrity?: string;
    keepalive?: boolean;
    method?: string;
    mode?: RequestMode;
    redirect?: RequestRedirect;
    referrer?: string;
    referrerPolicy?: ReferrerPolicy;
    window?: any;
}
```

The field is missing in definition, (but actually it exists). I tried installing `@types/node-fetch` but it didn't work. So, How to fix it?

File: package.json (hunk 1 above)
Comment (lihuanshuai) [reply package.json]: yes, I will modify it later

**Comment by KnisterPeter** [comment]
@lihuanshuai I will fix this later, then you can have a look at it.

File: src/provider/client.ts (hunk 2 above)
Comment (lihuanshuai) [reply src/provider/client.ts]: added in new commit

**Comment by KnisterPeter** [comment]
@lihuanshuai Nice. I'll fix the compile error and merge it afterwards. Thank for your contribution. :rocket:

**Pull request closed without merging**